File: day08/src/bin/part2.rs
use std::collections::{BTreeMap, HashMap};

fn process(input: &str) -> i64 {
    let mut lines = input
        .lines()
        .map(|line| line.trim())
        .filter(|line| !line.is_empty());

    let steps = lines.next().unwrap().chars().cycle();
    let map = parse_map(lines);

    least_common_multiple(
        map.keys()
            .filter(|key| key.chars().last().unwrap() == 'A')
            .map(|key| steps_to_end(key, &map, steps.clone())),
    )
}

type MapType<'a> = HashMap<&'a str, (&'a str, &'a str)>;

fn parse_map<'a>(lines: impl Iterator<Item = &'a str>) -> MapType<'a> {
    lines
        .map(|line| {
            let mut tokens = line.split('=');
            let key = tokens.next().unwrap().trim();
            let mut tokens = tokens
                .next()
                .unwrap()
                .trim()
                .trim_matches(&['(', ')'])
                .split(',');

            let left = tokens.next().unwrap().trim();
            let right = tokens.next().unwrap().trim();

            (key, (left, right))
        })
        .collect()
}

fn steps_to_end(start: &str, map: &MapType, steps: impl Iterator<Item = char>) -> i64 {
    let (mut count, mut node) = (0, start);
    for step in steps {
        if node.chars().last().unwrap() == 'Z' {
            return count;
        }

        count += 1;
        node = if step == 'L' {
            map.get(&node).unwrap().0
        } else if step == 'R' {
            map.get(&node).unwrap().1
        } else {
            panic!("Unexpected step direction: {step}")
        };
    }

    unreachable!()
}

fn least_common_multiple(items: impl Iterator<Item = i64>) -> i64 {
    let mut prime_counts = BTreeMap::new();

    for num in items {
        let factors = factorize(num);
        for (prime, count) in factors {
            prime_counts
                .entry(prime)
                .and_modify(|old_count| *old_count = count.max(*old_count))
                .or_insert(count);
        }
    }

    prime_counts
        .into_iter()
        .map(|(prime, count)| prime * count)
        .product()
}

fn factorize(mut num: i64) -> BTreeMap<i64, i64> {
    let primes = [
        2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89,
        97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181,
        191, 193, 197, 199, 211, 223, 227, 229, 233, 239, 241, 251, 257, 263, 269, 271, 277, 281,
        283, 293, 307, 311, 313, 317, 331, 337, 347, 349, 353, 359, 367, 373, 379, 383, 389, 397,
        401, 409, 419, 421, 431, 433, 439, 443, 449, 457, 461, 463, 467, 479, 487, 491, 499, 503,
        509, 521, 523, 541, 547, 557, 563, 569, 571, 577, 587, 593, 599, 601, 607, 613, 617, 619,
        631, 641, 643, 647, 653, 659, 661, 673, 677, 683, 691, 701, 709, 719, 727, 733, 739, 743,
        751, 757, 761, 769, 773, 787, 797, 809, 811, 821, 823, 827, 829, 839, 853, 857, 859, 863,
        877, 881, 883, 887, 907, 911, 919, 929, 937, 941, 947, 953, 967, 971, 977, 983, 991, 997,
    ];

    let mut result = BTreeMap::new();
    while num > 1 {
        for prime in primes {
            if num % prime == 0 {
                num /= prime;
                result
                    .entry(prime)
                    .and_modify(|count| *count += 1)
                    .or_insert(1);
            }
        }
    }
    result
}

fn main() {
    let input = include_str!("input.txt");
    println!("Result: {}", process(input));
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_process() {
        let input = "
            LR

            11A = (11B, XXX)
            11B = (XXX, 11Z)
            11Z = (11B, XXX)
            22A = (22B, XXX)
            22B = (22C, 22C)
            22C = (22Z, 22Z)
            22Z = (22B, 22B)
            XXX = (XXX, XXX)
        ";
        assert_eq!(process(input), 6);
    }
}

File: src/day04/part2.py
from collections import OrderedDict

def process(lines) -> int:
    cards = OrderedDict()
    for line in lines:
        key, nums = line.strip().split(':')
        key = int([key for key in key.split(' ') if key][1])

        win_nums, our_nums = nums.split('|')
        win_nums = [int(num) for num in win_nums.split(' ') if num]
        our_nums = [int(num) for num in our_nums.split(' ') if num]
        cards[key] = [len(set(win_nums) & set(our_nums)), 1]

    last_key = next(reversed(cards))
    for i in range(last_key + 1, last_key + 11):
        cards[i] = [0, 0]

    for key, (wins, count) in cards.items():
        for key in range(key + 1, key + wins + 1):
            cards[key][1] += count

    return sum(count for wins, count in cards.values())

def test():
    lines = """
    Card 1: 41 48 83 86 17 | 83 86  6 31 17  9 48 53
    Card 2: 13 32 20 16 61 | 61 30 68 82 17 32 24 19
    Card 3:  1 21 53 59 44 | 69 82 63 72 16 21 14  1
    Card 4: 41 92 73 84 69 | 59 84 76 51 58  5 54 83
    Card 5: 87 83 26 28 32 | 88 30 70 12 93 22 82 36
    Card 6: 31 18 13 56 72 | 74 77 10 23 35 67 36 11
    """.split('\n')

    lines = (line for line in lines if line.strip())
    assert process(lines) == 30, "Failed test"

if __name__ == '__main__':
    test()
    with open('input.txt', 'r') as file:
        print(f'Result: {process(file)}')

